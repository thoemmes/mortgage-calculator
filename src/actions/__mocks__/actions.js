

export const fetch = jest.fn();

// export const fetchEuribor = (url) => {

//     return fetch(url)
//         .then(response => {

//             if (response.ok) {
//                 //Promise with json data    
//                 return response.json().dataset.data[1];;
//             }

//             throw new Error('Network response was not ok.');
  
//         });

// }

export const fetchEuribor = (url) => {

    return fetch(url)
        .then(response => {

            if (response.ok) {
                //Promise with json data    
                return response.json().then(euriborData => euriborData.dataset.data[0][1]);
            }

            throw new Error('Network response was not ok.');

        });

}


// export const fetchEuribor = url => {
//     return new Promise((resolve, reject) => {
//         process.nextTick(() => {
//             if (url && url.localeCompare('https://www.quandl.com/api/v3/datasets/BOE/IUQAERB3.json?api_key=FfF1zN9ZYyTEitZzhEDJ') == 0) {
//                 resolve(-0.3299);
//             }
//             if (url && url.localeCompare('https://www.quandl.com/api/v3/datasets/BOE/IUQAERB3.json') == 0) {
//                 resolve({ok: false});
//             } else {
//                 reject(new Error('Network Error'));
//             }
//         });
//  });
// }



// export const fetchEuribor = url => {
//     return new Promise((resolve, reject) => {
//         console.log('P');
//         process.nextTick(() => {
//             console.log('PP');
//                 resolve(-0.3299);
//         });
//  });
// }