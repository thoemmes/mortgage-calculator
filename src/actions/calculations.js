export const calculateAmortisation = (mortgage, rate, interest) => {

    return yearlyAmortistaion(monthlyAmortisation(mortgage, rate, interest));
    
 }

// export const monthlyAmortisation = (mortgage, rate, interest) => {

//     let result = [];
    

//     let amortisation = 0;
//     let monthlyInterest;

//     do {
//         result.push({
//             mortgage,
//             amortisation
//         });

//         monthlyInterest = mortgage * interest / 12 * 0.01;
//         amortisation = rate - monthlyInterest;
//         mortgage = mortgage - amortisation;

//     } while (mortgage > 0 && amortisation > 0)

//     return result;
// }

const monthly= (mortgage, rate, interest) =>{
    
    let monthlyInterest = mortgage * interest / 12 * 0.01;
    let amortisation = rate - monthlyInterest;
        
    return {
        mortgage : mortgage - amortisation,
        amortisation
    };   
}

const amortised = (payment) => {
    return payment.mortgage <= 0 || payment.amortisation < 0;
}

export const monthlyAmortisation = (mortgage, rate, interest) => {

    let result = [];
    
    let payment =  monthly(mortgage, rate, interest);

    while(!amortised(payment)){
        result.push(payment);
        payment= monthly(payment.mortgage, rate, interest);
        
    } 
    
    return result;
}


export const yearlyAmortistaion = monthly => {

    let result = [];
    //monthly.splice(0,1);
    while (monthly.length > 0){
        let year = monthly.splice(0, 12);

        result.push(year.reduce((acc, m) => {
            acc.mortgage = m.mortgage;
            acc.amortisation += m.amortisation;
            acc.total= acc.mortgage + acc.amortisation;
            return acc;
        }, { mortgage: 0, amortisation: 0, total: 0 })
        );

    } 
    
    return result;
}