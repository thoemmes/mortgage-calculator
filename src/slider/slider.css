

.slider {
  flex-basis: 100%;
  border-style: solid;
  border-width: 0px 1rem;
  border-color: #ff7f27; 
  background-image: url("slider.png");
  
  background-size: 100% 100%;
  padding: 0.3rem 1rem 0.3rem 0rem;
  font-size: 1rem;
  text-align: left;

  user-select: none;
 
}

.outer_handle{
padding: 0px;
margin-left: -10px;
}
.handle{
user-select: none;
background-color: gray;
padding: 0px 0.5rem;

}

.value{
  flex-basis: 4rem;
  margin-left: 0.5rem;
  margin-right: 0.5rem; 
  text-align: center;
  user-select: none;
  
}