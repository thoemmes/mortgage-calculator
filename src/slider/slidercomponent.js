import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Slider from './slider'


const mapStateToProps = (state) => {
  return {
    sliderPosition: state.ui.mortgageSlider.position,
    cursor: (state.ui.mortgageSlider.isMoving ? 'e-resize' : 'default')
  }
}


const setMortgageSliderPosition = (pos) => {
  return {
    type: 'SET_MORTGAGE_SLIDER',
    position: pos
  }
}


const setMortgageSliderMove = (isMoving) => {
  return {
    type: 'MORTGAGE_SLIDER_MOVING',
    isMoving
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    setSliderPosition: (v) => {
      dispatch(setMortgageSliderPosition(v));
    },
    mouseDown: () => {
      dispatch(setMortgageSliderMove(true));
    },
    mouseUp: () => {
      dispatch(setMortgageSliderMove(false));

    }
  }
}

const SliderComponent = connect(mapStateToProps, mapDispatchToProps)(Slider);

// SliderComponent.propTypes = {
//   setSliderPosition: PropTypes.func.isRequired,
//   //   mouseDown: PropTypes.func.isRequired,
//   //   mouseUp: PropTypes.func.isRequired,
//   //   sliderPosition: PropTypes.number.isRequired,
//   cursor: PropTypes.string.isRequired
// };

export default SliderComponent;

