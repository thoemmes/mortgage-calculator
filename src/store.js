import { createStore, combineReducers, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';


// const ui = (state, action) => {
//     console.log(state);
//     switch (action.type) {
//         case 'SET_MORTGAGE_SLIDER':
//         case 'MORTGAGE_SLIDER_MOVING':
//             return {
//                 ...state,
//                 mortgageSlider: mortgageSlider(state.mortgageSlider, action)
//             };
//         case 'SET_AMORTISATION_SLIDER':
//         case 'AMORTISATION_SLIDER_MOVING': 
//             return{
//                 ...state,
//                 amortisationSlider: amortisationSlider(state.amortisationSlider, action)
//             }; 

//         default:
//             return { 
//                 mortgageSlider : mortgageSlider(state, action),
//                 amortisationSlider : amortisationSlider(state, action)
//             };
//     }
// }



const mortgageSlider = (state = {
    position: 0.5,
    isMoving: false
}, action) => {
    switch (action.type) {
        case 'SET_MORTGAGE_SLIDER':
            return {
                ...state,
                position: action.position,
            };
        case 'MORTGAGE_SLIDER_MOVING':
            return {
                ...state,
                isMoving: action.isMoving
            }
        default:
            return state;
    }
}

const amortisationSlider = (state = {
    position: 0.5,
    isMoving: false
}, action) => {
    switch (action.type) {
        case 'SET_AMORTISATION_SLIDER':
            return {
                ...state,
                position: action.position,
            };
        case 'AMORTISATION_SLIDER_MOVING':
            return {
                ...state,
                isMoving: action.isMoving
            }
        default:
            return state;
    }
}


const interestSlider = (state = {
    position: 0.5,
    isMoving: false
}, action) => {
    switch (action.type) {
        case 'SET_INTEREST_SLIDER':
            return {
                ...state,
                position: action.position,
            };
        case 'INTEREST_SLIDER_MOVING':
            return {
                ...state,
                isMoving: action.isMoving
            }
        default:
            return state;
    }
}


const ui = combineReducers({mortgageSlider, amortisationSlider, interestSlider});

const mortgageMax = (state = 500000
    , action) => {
    switch (action.type) {

        case 'SET_MORTGAGE_MAX': {
            return action.max;

        }
        default:
            return state;
    }
}

const amortisationMax = (state = 2000
    , action) => {
    switch (action.type) {

        case 'SET_AMORTISATION_MAX': {
            return action.max;

        }
        default:
            return state;
    }
}

const interestMax = (state = 5
    , action) => {
    switch (action.type) {

        case 'SET_INTEREST_MAX': {
            return action.max;

        }
        default:
            return state;
    }
}


// const interest = (state = 0.02, action) => {
//     switch (action.type) {
//         case 'SET_INTEREST':
//             return action.interest;
//         default:
//             return state;
//     }
// }

// const amortisation = (state = 0.02, action) => {
//     switch (action.type) {
//         case 'SET_AMORTISATION':
//             return action.amortisation;
//         default:
//             return state;
//     }
// }


export const reducers = combineReducers({
    ui,
    mortgageMax,
    //interest,
    amortisationMax,
    interestMax
});

const store = createStore(reducers, applyMiddleware(thunkMiddleware));


export default store;




// const mortgage = (mortgage: 50000, action) => {
//     switch (action.type) {
//         case 'MORTGAGE':
//             return {
//                 ...state,
//                 mortgage: action.amount
//             };
//     }

//     return state;
// }

// const mortgageSliderPosition = (state = 0.5, action) => {
//     switch (action.type) {
//         case 'SET_MORTGAGE_SLIDER':
//             return action.position;
//         default:
//             return state;
//     }
// }

// const mortgageSliderCursor = (state = 'default', action) => {
//     switch (action.type) {
//         case 'SET_CURSOR':
//             return action.cursor;
//         default:
//             return state;
//     }
// }

// const mortgageSliderMove = (state = 'default', action) => {
//     switch (action.type) {
//         case 'MORTGAGE_SLIDER_MOVE':
//             return action.move ? 'e-resize' : 'default';
//         default:
//             return state;
//     }
// }
