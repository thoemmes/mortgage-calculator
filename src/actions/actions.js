import { createSelector } from 'reselect';
import { calculateAmortisation } from './calculations';

export const setMortgageSliderPosition = (pos) => {
    return {
        type: 'SET_MORTGAGE_SLIDER',
        position: pos
    }
}


export const setMortgageSliderMove = (isMoving) => {
    return {
        type: 'MORTGAGE_SLIDER_MOVING',
        isMoving
    }
}



export const setMortgageMax = (max) => {
    return {
        type: 'SET_MORTGAGE_MAX',
        max
    }
}


export const setAmortisationSliderPosition = (pos) => {
    return {
        type: 'SET_AMORTISATION_SLIDER',
        position: pos
    }
}


export const setAmortisationSliderMove = (isMoving) => {
    return {
        type: 'AMORTISATION_SLIDER_MOVING',
        isMoving
    }
}



export const setAmortisationMax = (max) => {
    return {
        type: 'SET_AMORTISATION_MAX',
        max
    }
}


export const setInterestSliderPosition = (pos) => {
    return {
        type: 'SET_INTEREST_SLIDER',
        position: pos
    }
}

export const setInterestMax = (max) => {
    return {
        type: 'SET_INTEREST_MAX',
        max
    }
}


export const setInterestSliderMove = (isMoving) => {
    return {
        type: 'INTEREST_SLIDER_MOVING',
        isMoving
    }
}



export const setInterest = (interest) => {
    return {
        type: 'SET_INTEREST',
        interest
    }
}


export const REQUEST_EURIBOR = 'REQUEST_EURIBOR'

export const requestEuribor = (period) => {
    return {
        type: REQUEST_EURIBOR,
        period
    }
}


export const RECEIVE_EURIBOR = 'RECEIVE_EURIBOR'

export const receiveEuribor = (response) => {
    return {
        type: RECEIVE_EURIBOR,
        response
    }
}


// export const fetchEuribor = (period) => {

//     // Thunk middleware knows how to handle functions.
//     // It passes the dispatch method as an argument to the function,
//     // thus making it able to dispatch actions itself.

//     //return (dispatch) => {

//     // First dispatch: the app state is updated to inform
//     // that the API call is starting.

//     //dispatch(requestPosts(subreddit))

//     // The function called by the thunk middleware can return a value,
//     // that is passed on as the return value of the dispatch method.

//     // In this case, we return a promise to wait for.
//     // This is not required by thunk middleware, but it is convenient for us.

//     //https://www.quandl.com/api/v3/datasets/BOE/IUQAERB3.json?api_key=FfF1zN9ZYyTEitZzhEDJ

//     return fetch(`https://www.quandl.com/api/v3/datasets/BOE/IUQAERB3.json?api_key=FfF1zN9ZYyTEitZzhEDJ`)
//         .then(response => {

//             if (response.ok) {
//                 //Promise with json data    
//                 return response.json();
//             }

//             throw new Error('Network response was not ok.');
//             //console.log('fetch');
//             //console.log(response);

//             // We can dispatch many times!
//             // Here, we update the app state with the results of the API call.

//             //dispatch(receiveEuribor(response))
//             //dispatch(setInterest(0.1));
//         })

//     // In a real world app, you also want to
//     // catch any error in the network call.
//     //}
// }


export const fetchEuribor = (url) => {

    return fetch(url)
        .then(response => {

            if (response.ok) {
                //Promise with json data    
                return response.json().then(euriborData => euriborData.dataset.data[0][1]);
            }

            throw new Error('Network response was not ok.');

        });

}



export const fetchEuriborX = (fetch, url) => {

    return fetch(url); //.then(response => {

    //return new Promise( () => -0.3299);
    // if (response.ok) {
    //     //Promise with json data    
    //     return response.json();
    // }

    // throw new Error('Network response was not ok.');
    //});

}





const sliderPosition = state => state.ui.mortgageSlider.position;
const mortgageMaximum = state => state.mortgageMax;

const isMoving = state => state.ui.mortgageSlider.isMoving;


export const mortgageSelector = createSelector([sliderPosition, mortgageMaximum], (sliderPos, max) => {
    let raw = sliderPos * max;
    if (raw < 1000) {
        return raw;
    }
    return Math.round(sliderPos * max / 1000) * 1000;
});


export const cursorSelector = createSelector([isMoving], (isMoving) => isMoving ? 'e-resize' : 'default');



const amortisationSliderPosition = state => state.ui.amortisationSlider.position;
const amortisationMaximum = state => state.amortisationMax;

const amortisationIsMoving = state => state.ui.amortisationSlider.isMoving;

export const amortisationSelector = createSelector([amortisationSliderPosition, amortisationMaximum], (sliderPos, max) => {
    let raw = sliderPos * max;
    if (raw < 100) {
        return raw;
    }
    return Math.round(sliderPos * max / 100) * 100;

});


export const amortisationCursorSelector = createSelector([amortisationIsMoving], (isMoving) => isMoving ? 'e-resize' : 'default');


const interestSliderPosition = state => state.ui.interestSlider.position;
const interestMaximum = state => state.interestMax;

const interestIsMoving = state => state.ui.interestSlider.isMoving;

export const interestSelector = createSelector([interestSliderPosition, interestMaximum], (sliderPos, max) => {
    return Math.round(sliderPos * max * 10) / 10;
});


export const interestCursorSelector = createSelector([interestIsMoving], (isMoving) => isMoving ? 'e-resize' : 'default');



export const mortgage = state => state.ui.mortgageSlider.position * state.mortgageMax
export const amortisation = state => state.ui.amortisationSlider.position * state.amortisationMax
export const interestRate = state => state.ui.interestSlider.position * state.interestMax

export const ratesSelector = createSelector([mortgage, amortisation, interestRate], (mortgage, amortisation, interestRate) => {

    let payments = calculateAmortisation(mortgage, amortisation, interestRate);
    payments.unshift({ mortgage, amortisation: 0, total: mortgage });

    return payments;

});



