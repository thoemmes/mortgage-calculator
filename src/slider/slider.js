import React from 'react';
import './slider.css';

//const isLeftButtonPressed = ev => ev.buttons === 1;
const sliderInMoveMode = cursor => cursor !== 'default';


const calculateRelativePosition = (sliderPos, rect) => {

  let value = (sliderPos - rect.left) / (rect.width);

  if (value > 1) {
    value = 1;
  } else if (value < 0) {
    value = 0;
  }
  return value;
}

// A horizontal slider component. The slider can be "dragged" by pressing the left mouse button
// The slider is positioned via the margin-left css property. The Position prop is manipulated by calling
// the setSliderPosition function, sliderPosition is in the range of [0,1]
const Slider = ({ setSliderPosition, mouseDown, mouseUp, sliderPosition, cursor, rect }) => {

  const mouseMove = (ev) => {
    //if (ev.buttons !== 0 && rect.width !== 0) {
    if (sliderInMoveMode(cursor) && rect.width !== 0) {

      let newPos = calculateRelativePosition(ev.clientX, rect);
      setSliderPosition(newPos);

    }
  }
  
  return (
    <div className="slider"
      
      onMouseMove={mouseMove}
      //By calling mouseMove it is assured that the user can reach the slider limits
      // regardless of screen resolution issues
      onMouseLeave={(e) => { mouseMove(e); mouseUp(e); }}
      //onMouseLeave={(e) => { mouseMove(e); }}
      onMouseDown={mouseDown}
      onMouseUp={mouseUp}
      ref={el => { if (el && !rect) rect = el.getBoundingClientRect(); }}
      style={{
        cursor: cursor
      }}
    >
      <div className="outer_handle"
        draggable="false"
        style={{
          marginLeft: sliderPosition * 100 + '%',
          cursor: cursor
        }}><span className="handle">IIII</span></div>
    </div>
  )
};



export default Slider;