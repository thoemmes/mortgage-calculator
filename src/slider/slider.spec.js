import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import Slider from './slider';
import './slider.css';
import SliderComponent from './slidercomponent';
import { setMortgageSliderPosition } from '../actions/actions';
import { Provider } from 'react-redux';
import { reducers } from '../store';
import { createStore } from 'redux';

describe('Slider', () => {

  const initialState = {
    mortgageSlider: {
      sliderPosition: 0.5,
      isMoving: true
    }
  };
  const mockStore = configureStore();
  let store;
  const el = document.createElement('div');
  beforeEach(() => {
    //store = mockStore(initialState);
    store = createStore(reducers);

  })

  it('renders without crashing', () => {
    shallow(<Slider />);
  });

  it('renders default cursor', () => {
    const component = renderer.create(<Slider cursor='default' />);
    //let handle= slider.find('.outer_handle');

    const slider = component.toJSON();

    expect(slider).toMatchSnapshot();
    expect(slider.props.style.cursor).toEqual('default');
  });


  it('renders starts with position 0.5', () => {
    const component = renderer.create(<Slider cursor='default' sliderPosition='0.5' />);

    const slider = component.toJSON();
    const handle = slider.children[0];

    expect(handle.props.style.marginLeft).toEqual('50%');
  });

  it('moves slider with mouse move and button left', () => {


    // const component = mount(<Provider store={store}><SliderComponent rect={{ left: 0, width: 100 }}/></Provider> );

    const component = mount(<SliderComponent store={store} rect={{ left: 0, width: 100 }} />);

    component.simulate('mousedown');
    component.simulate('mousemove', { buttons: 1, clientX: 90 });
    /*  let action = store.getActions();
      expect(action[0].type).toBe("SET_MORTGAGE_SLIDER");
      expect(action[0].position).toEqual(0.9);
      */
    //store.dispatch(setMortgageSliderPosition(0.9));

    let handle = component.find('.outer_handle');
    expect(handle.prop('style').marginLeft).toEqual('90%');

  });


  it(' should not move with no mouse buttons pressed', () => {


    const component = mount(<SliderComponent store={store} rect={{ left: 0, width: 100 }} />);

    component.simulate('mousemove', { buttons: 0, clientX: 90 });

    let handle = component.find('.outer_handle');
    expect(handle.prop('style').marginLeft).toEqual('50%');

  });

 it('changes cursor to e-resize on mouse down', () => {


    const sliderComponent = mount(<SliderComponent store={store} rect={{ left: 0, width: 100 }} />);

    sliderComponent.simulate('mousedown');
    
    let handle = sliderComponent.find('.outer_handle');
    expect(handle.prop('style').cursor).toEqual('e-resize');
    let slider = sliderComponent.find('.outer_handle');
    expect(slider.prop('style').cursor).toEqual('e-resize');
  });





 it('changes cursor to default on mouse up', () => {


    const sliderComponent = mount(<SliderComponent store={store} rect={{ left: 0, width: 100 }} />);

    sliderComponent.simulate('mousedown');
    
    const handle = sliderComponent.find('.outer_handle');
    const slider = sliderComponent.find('.slider');

    expect(handle.prop('style').cursor).toEqual('e-resize');

    sliderComponent.simulate('mouseup');
    expect(handle.prop('style').cursor).toEqual('default');
    expect(slider.prop('style').cursor).toEqual('default');

  });


})