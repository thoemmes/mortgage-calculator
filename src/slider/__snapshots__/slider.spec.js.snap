// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider renders default cursor 1`] = `
<div
  className="slider"
  onMouseDown={undefined}
  onMouseLeave={[Function]}
  onMouseMove={[Function]}
  onMouseUp={undefined}
  style={
    Object {
      "cursor": "default",
    }
  }
>
  <div
    className="outer_handle"
    draggable="false"
    style={
      Object {
        "cursor": "default",
        "marginLeft": "NaN%",
      }
    }
  >
    <span
      className="handle"
    >
      IIII
    </span>
  </div>
</div>
`;
